### ISSUE TITLE
Bug: `import/no-unused-modules` ESLint couldn't find a non-standard name configuration file

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#16787]
### Environment

Node version: v18.12.1
npm version: 9.2.0
Local ESLint version: 8.32.0
Global ESLint version: N/A
Operating System: Ubuntu 22.04.1 LTS

### What parser are you using?

Default (Espree)

### What did you do?

**package.json**:

```JSON
{
    "name": "testcase",
    "version": "1.0.0",
    "type": "module",
    "private": true,
    "dependencies": {
        "eslint-plugin-import": "2.27.4",
        "eslint": "8.32.0"
    }
}
```

**my_eslint.json**:

```JSON
{
    "plugins": ["import"],
    "parserOptions": {
        "sourceType": "module",
        "ecmaVersion": 2022
    },

    "rules": {
        "import/no-unused-modules": [2, { "unusedExports": true }]
    }
}
```

**index.js:**

```js
console.log("foo");
```

Run this command: **`npx eslint -c my_eslint.json --debug index.js`**

If I rename the configuration file to `.eslintrc.json`, there is no more problem. The error occurs when using a non-standard ESLint configuration file name (different from `.eslintrc.*`).

### What did you expect to happen?

No error reported.

### What actually happened?

```
  eslint:cli CLI args: [ '-c', 'my_eslint.json', '--debug', 'index.js' ] +0ms
  eslint:cli Using flat config? false +4ms
  eslint:cli Running on files +3ms
  eslintrc:config-array-factory Loading JSON config file: /home/regseb/testcase/package.json +0ms
  eslintrc:config-array-factory Loading JSON config file: /home/regseb/testcase/my_eslint.json +0ms
  eslintrc:config-array-factory Loading plugin "import" from /home/regseb/testcase/my_eslint.json +1ms
  eslintrc:config-array-factory Loaded: eslint-plugin-import@2.27.4 (/home/regseb/testcase/node_modules/eslint-plugin-import/lib/index.js) +0ms
  eslintrc:config-array-factory Plugin /home/regseb/testcase/node_modules/eslint-plugin-import/lib/index.js loaded in: 71ms +71ms
  eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/home/regseb/testcase', loose: false } ] +0ms
  eslintrc:ignore-pattern   processed: { basePath: '/home/regseb/testcase', patterns: [ '/**/node_modules/*' ] } +1ms
  eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/home/regseb/testcase', loose: false } ] +1ms
  eslintrc:ignore-pattern   processed: { basePath: '/home/regseb/testcase', patterns: [ '/**/node_modules/*' ] } +0ms
  eslint:file-enumerator Start to iterate files: [ 'index.js' ] +0ms
  eslint:file-enumerator File: /home/regseb/testcase/index.js +0ms
  eslintrc:cascading-config-array-factory Load config files for /home/regseb/testcase. +0ms
  eslintrc:cascading-config-array-factory No cache found: /home/regseb/testcase. +0ms
  eslintrc:config-array-factory Loading package.json config file: /home/regseb/testcase/package.json +5ms
  eslintrc:config-array-factory Loading JSON config file: /home/regseb/testcase/package.json +0ms
  eslintrc:config-array-factory Error reading package.json file: /home/regseb/testcase/package.json +0ms
  eslintrc:config-array-factory Config file not found on /home/regseb/testcase +0ms
  eslintrc:cascading-config-array-factory No cache found: /home/regseb. +1ms
  eslintrc:cascading-config-array-factory Stop traversing because of considered root. +0ms
  eslintrc:cascading-config-array-factory Configuration was determined: ConfigArray(2) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/home/regseb/testcase', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '--config', filePath: '/home/regseb/testcase/my_eslint.json', criteria: null, env: undefined, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: undefined, parserOptions: { sourceType: 'module', ecmaVersion: 2022 }, plugins: { import: [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'import/no-unused-modules': [Array] }, settings: undefined } ] on /home/regseb/testcase +3ms
  eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/home/regseb/testcase', loose: false } ] +6ms
  eslintrc:ignore-pattern   processed: { basePath: '/home/regseb/testcase', patterns: [ '/**/node_modules/*' ] } +0ms
  eslintrc:ignore-pattern Check {
  filePath: '/home/regseb/testcase/index.js',
  dot: false,
  relativePath: 'index.js',
  result: false
} +1ms
  eslint:cli-engine Lint /home/regseb/testcase/index.js +0ms
  eslint:linter Linting code for /home/regseb/testcase/index.js (pass 1) +0ms
  eslint:linter Verify +0ms
  eslint:linter With ConfigArray: /home/regseb/testcase/index.js +0ms
  eslint:linter Parsing: /home/regseb/testcase/index.js +1ms
  eslint:linter Parsing successful: /home/regseb/testcase/index.js +5ms
  eslint:linter Scope analysis: /home/regseb/testcase/index.js +0ms
  eslint:linter Scope analysis successful: /home/regseb/testcase/index.js +1ms
  eslintrc:config-array-factory Loading JSON config file: /home/regseb/testcase/package.json +14ms
  eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/home/regseb/testcase', loose: false } ] +9ms
  eslintrc:ignore-pattern   processed: { basePath: '/home/regseb/testcase', patterns: [ '/**/node_modules/*' ] } +0ms
  eslint:file-enumerator Start to iterate files: [ '/home/regseb/testcase' ] +15ms
  eslint:file-enumerator Directory: /home/regseb/testcase +0ms
  eslint:file-enumerator Enter the directory: /home/regseb/testcase +0ms
  eslintrc:cascading-config-array-factory Load config files for /home/regseb/testcase. +12ms
  eslintrc:cascading-config-array-factory No cache found: /home/regseb/testcase. +0ms
  eslintrc:config-array-factory Loading package.json config file: /home/regseb/testcase/package.json +1ms
  eslintrc:config-array-factory Loading JSON config file: /home/regseb/testcase/package.json +0ms
  eslintrc:config-array-factory Error reading package.json file: /home/regseb/testcase/package.json +0ms
  eslintrc:config-array-factory Config file not found on /home/regseb/testcase +0ms
  eslintrc:cascading-config-array-factory No cache found: /home/regseb. +0ms
  eslintrc:cascading-config-array-factory Stop traversing because of considered root. +0ms
  eslintrc:cascading-config-array-factory Loading the config file of the home directory: /home/regseb +0ms
  eslintrc:config-array-factory Config file not found on /home/regseb +0ms
  eslintrc:cascading-config-array-factory Configuration was determined: ConfigArray(1) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/home/regseb/testcase', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined } ] on /home/regseb/testcase +0ms
  eslintrc:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/home/regseb/testcase', loose: false } ] +1ms
  eslintrc:ignore-pattern   processed: { basePath: '/home/regseb/testcase', patterns: [ '/**/node_modules/*' ] } +1ms
  eslintrc:ignore-pattern Check {
  filePath: '/home/regseb/testcase/index.js',
  dot: false,
  relativePath: 'index.js',
  result: false
} +0ms
  eslint:file-enumerator Yield: index.js +2ms
  eslintrc:cascading-config-array-factory Load config files for /home/regseb/testcase. +1ms
  eslintrc:cascading-config-array-factory Cache hit: /home/regseb/testcase. +0ms
  eslint:linter An error occurred while traversing +4ms
  eslint:linter Filename: /home/regseb/testcase/index.js +0ms
  eslint:linter Parser Options: {
  sourceType: 'module',
  ecmaVersion: 13,
  ecmaFeatures: { globalReturn: false }
} +0ms
  eslint:linter Parser Path: espree +0ms
  eslint:linter Settings: {} +0ms

Oops! Something went wrong! :(

ESLint: 8.32.0

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config

ESLint looked for configuration files in /home/regseb/testcase and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://eslint.org/chat/help
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

- I also had the problem with ESLint 8.32.0.
- I have a configuration with about 500 rules (from ESLint and 9 plugins) and [`import/no-unused-modules`](https://github.com/import-js/eslint-plugin-import/blob/v2.27.4/src/rules/no-unused-modules.js) is the only rule with this problem.
- I opened a ticket at eslint-plugin-import, but _this plugin doesn't do anything with eslint config._ https://github.com/import-js/eslint-plugin-import/issues/2678

**Comment by mdjermanovic** [comment]
Hi @regseb, thanks for the issue! 

> * I opened a ticket at eslint-plugin-import, but _this plugin doesn't do anything with eslint config._ [[no-unused-modules] Error with ESLint configuration non-standard name import-js/eslint-plugin-import#2678](https://github.com/import-js/eslint-plugin-import/issues/2678)

The `import/no-unused-modules` rule indirectly loads eslint configs by [using](https://github.com/import-js/eslint-plugin-import/blob/0778b0390ec18b9fdc1acfd45d10bf1b614d82ff/src/rules/no-unused-modules.js#L21) eslint's internal module `FileEnumerator`. This module isn't intended to be used outside eslint package. When used by eslint, `FileEnumerator` receives full execution context that includes processed CLI options. The error you're getting is caused by the rule not providing the context to the `FileEnumerator` [instance](https://github.com/import-js/eslint-plugin-import/blob/0778b0390ec18b9fdc1acfd45d10bf1b614d82ff/src/rules/no-unused-modules.js#L57-L59) it's using, so it works as if `-c my_eslint.json` is not specified and therefore looks for `.eslintrc.*` config files. We can discuss in https://github.com/eslint/eslint/issues/13481 how this rule could work with the new config system (cc @ljharb), but until then it seems incompatible with the `-c` usage.

**Comment by ljharb** [comment]
Ah, thanks for clarifying.

**Comment by ljharb** [comment]
@mdjermanovic does that mean if we passed the context to the FileEnumerator instance, it would Just Work with the legacy config system?

**Comment by mdjermanovic** [comment]
It might work, but I don't think that rules have access to all the data that ESLint passes to FileEnumerator.

https://github.com/eslint/eslint/blob/d8c8ede088e1f4a82c9b6b5c2772af268b9161aa/lib/cli-engine/cli-engine.js#L606-L628

**Comment by ljharb** [comment]
ah, and that's not something eslint exposes to rules?

**Comment by mdjermanovic** [comment]
No, and `FileEnumerator` was not intended to be used by rules or in any way other than internally by other ESLint modules. I think the best course of action would be to open an issue describing what this rule needs from ESLint and how that could be implemented in the new config system.

**Comment by ljharb** [comment]
I'll try to find time to do that. Unfortunately, that won't address issues with existing users, and I'm highly disinclined to make a breaking change.
